src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,404 @@
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Added, cancellationToken);
+        }
+
+        Task IBlockListener.BlockRemovingAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Removing, cancellationToken);
+        }
+
+        async Task<IEnumerable<Rule>> ITransactionConfirmationWatcherHandler<Rule>.CreateContextsAsync(Transaction tx, CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterReadLock();
+
+            try
+            {
+                if (this.timers.TryGetValue(tx.GetHash(), out var txTimers) && txTimers.Count > 0)
+                {
+                    var contexts = new Collection<Rule>();
+
+                    foreach (var timer in txTimers)
+                    {
+                        var rule = await this.ruleRepository.GetAsync(timer.Key, cancellationToken);
+                        contexts.Add(rule);
+                    }
+
+                    return contexts;
+                }
+            }
+            finally
+            {
+                this.timerLock.ExitReadLock();
+            }
+
+            return Enumerable.Empty<Rule>();
+        }
+
+        async Task<bool> IConfirmationWatcherHandler<TransactionWatch<Rule>, Rule>.ConfirmationUpdateAsync(TransactionWatch<Rule> watch, int confirmation, ConfirmationType type, CancellationToken cancellationToken)
+        {
+            switch (type)
+            {
+                case ConfirmationType.Confirmed:
+                    if (confirmation >= watch.Context.Confirmation)
+                    {
+                        await ConfirmAsync(watch, cancellationToken);
+                        return true;
+                    }
+                    break;
+                case ConfirmationType.Unconfirming:
+                    break;
+                default:
+                    throw new NotSupportedException($"{type} is not supported.");
+            }
+
+            return false;
+        }
+
+        Task<IEnumerable<TransactionWatch<Rule>>> IConfirmationWatcherHandler<TransactionWatch<Rule>, Rule>.GetCurrentWatchesAsync(CancellationToken cancellationToken)
+        {
+            return this.watchRepository.ListAsync(TransactionConfirmationWatchingWatchStatus.Pending, cancellationToken);
+        }
+
+        async Task IWatcherHandler<TransactionWatch<Rule>, Rule>.AddWatchesAsync(IEnumerable<TransactionWatch<Rule>> watches, CancellationToken cancellationToken)
+        {
+            if (watches == null)
+            {
+                throw new ArgumentNullException(nameof(watches));
+            }
+
+            foreach (var watch in watches) // lgtm [cs/linq/missed-where]
+            {
+                if (await StopTimer(watch.Context))
+                {
+                    await this.watchRepository.AddAsync(watch, cancellationToken);
```

Comment:
Seem like we forgot to assign current watch for rule here.

Reply:
Currently `AddAsync` do that but I have moved it to rules repository